fq_test.go

```diff
@@ -0,0 +1,275 @@
+}
+
+func genFlow(fq *FQScheduler, desc *flowDesc, key uint64, done_wg *sync.WaitGroup) {
+	for i, t := uint64(1), uint64(0); t < desc.ftotal; i++ {
+		it := new(Packet)
+		it.key = key
+		if desc.imin == desc.imax {
+			it.size = desc.imax
+		} else {
+			it.size = desc.imin + uint64(rand.Int63n(int64(desc.imax-desc.imin)))
+		}
+		if t+it.size > desc.ftotal {
+			it.size = desc.ftotal - t
+		}
+		t += it.size
+		it.seq = i
+		// new packet
+		fq.Enqueue(it)
+	}
+	(*done_wg).Done()
+}
+
+func consumeQueue(t *testing.T, fq *FQScheduler, descs []flowDesc) (float64, error) {
+	active := make(map[uint64]bool)
+	var total uint64
+	acnt := make(map[uint64]uint64)
+	cnt := make(map[uint64]uint64)
+	seqs := make(map[uint64]uint64)
+
+	var wsum uint64
+	for range descs {
+		wsum += uint64(1)
+	}
+
+	// for !fq.seen {
+
+	// }
+	// TODO(aaron-prindle) remove this
+	time.Sleep(1 * time.Second)
+	for i, ok := fq.processround(); ok; i, ok = fq.processround() {
+		time.Sleep(time.Microsecond) // Simulate constrained bandwidth
+		it := i
+		seq := seqs[it.key]
+		if seq+1 != it.seq {
+			return 0, fmt.Errorf("Packet for flow %d came out of queue out-of-order: expected %d, got %d", it.key, seq+1, it.seq)
+		}
+		seqs[it.key] = it.seq
+
+		// set the flow this item is a part of to active
+		if cnt[it.key] == 0 {
+			active[it.key] = true
+		}
+		cnt[it.key] += it.size
+
+		// if # of active flows is equal to the # of total flows, add to total
+		// we are correctly taking a bit from each
+		if len(active) == len(descs) {
+			acnt[it.key] += it.size
+			total += it.size
+		}
+
+		// if all items have been processed from the flow, remove it from active
+		if cnt[it.key] == descs[it.key].ftotal {
+			delete(active, it.key)
+		}
+	}
+
+	var variance float64
+	for key := uint64(0); key < uint64(len(descs)); key++ {
+		if total == 0 {
+			t.Fatalf("expected 'total' to be nonzero")
+		}
+		// idealPercent = total-all-active/len(flows) / total-all-active
+		// idealPercent = ~((10000/10) / 10000) * 100 = ~10
+		// "how many bytes/requests you expect for this flow - all-active"
+		descs[key].idealPercent = (((float64(total)) / float64(wsum)) / float64(total)) * 100
+
+		// actualPercent = requests-for-this-flow-all-active / total-reqs
+		// "how many bytes/requests you got for this flow - all-active"
+		descs[key].actualPercent = (float64(acnt[key]) / float64(total)) * 100
+
+		x := descs[key].idealPercent - descs[key].actualPercent
+		x *= x
+		variance += x
+	}
+	fmt.Println(descs)
+
+	stdDev := math.Sqrt(variance)
+	return stdDev, nil
+}
+
+func TestSingleFlow(t *testing.T) {
+	runtime.GOMAXPROCS(runtime.NumCPU())
+	queues := initQueues(1, 1)
+	fq := newFQScheduler(queues)
+
+	go func() {
+		for i := 1; i < 10000; i++ {
+			it := &Packet{}
+			it.key = 1
+			it.size = uint64(rand.Int63n(10) + 1)
+			it.seq = uint64(i)
+			fq.Enqueue(it)
+		}
+	}()
+
+	var seq uint64
+	hasEntered := false
+	for hasEntered {
+		for it, ok := fq.Dequeue(); ok; it, ok = fq.Dequeue() {
```

Comment:
It's a little disconcerting to see `Dequeue()` used in one test and `processround()` used in the other tests.

Since the virtual clock is not used in dispatching (the virtual clock is only read when a packet arrives), and the difference between `processround()` and `Dequeue()` is mainly in whether the virtual clock is advanced or not, and it looks like your intent is to have all the packet arrivals happen before any dispatching starts, I think you could pick either one and use it in all tests.

Reply:
Sorry it is definitely unclear but some of these were out of alignment due to my changing of the code.  I was primarily using TestMultiFlow.  I will fix these differences